File: core/phone_utils.py
"""Vietnam phone extraction shared by Facebook comments and CRM leads."""

import re


PHONE_CANDIDATE_RE = re.compile(r'(?<!\d)(?:(?:\+|00)?84|0)(?:[\s.\-()]?\d){8,10}(?!\d)')
VIETNAM_MOBILE_RE = re.compile(r'^0(?:3|5|7|8|9)\d{8}$')
VIETNAM_LANDLINE_RE = re.compile(r'^02\d{8,9}$')


def normalize_phone(raw: str) -> str:
    digits = re.sub(r'\D', '', raw or '')
    if digits.startswith('0084'):
        digits = f'0{digits[4:]}'
    elif digits.startswith('84'):
        digits = f'0{digits[2:]}'
    if VIETNAM_MOBILE_RE.fullmatch(digits) or VIETNAM_LANDLINE_RE.fullmatch(digits):
        return digits
    return ''


def extract_phones(text: str) -> list[str]:
    seen: set[str] = set()
    phones: list[str] = []
    for match in PHONE_CANDIDATE_RE.finditer(text or ''):
        phone = normalize_phone(match.group())
        if phone and phone not in seen:
            seen.add(phone)
            phones.append(phone)
    return phones

File: web/src/lib/phone-utils.ts
const PHONE_RE = /(?<!\d)(?:(?:\+|00)?84|0)(?:[\s.\-()]?\d){8,10}(?!\d)/g;
const VIETNAM_MOBILE_RE = /^0(?:3|5|7|8|9)\d{8}$/;
const VIETNAM_LANDLINE_RE = /^02\d{8,9}$/;

export function normalizePhone(raw: string): string {
  let digits = (raw || '').replace(/\D/g, '');
  if (digits.startsWith('0084')) digits = `0${digits.slice(4)}`;
  else if (digits.startsWith('84')) digits = `0${digits.slice(2)}`;
  if (VIETNAM_MOBILE_RE.test(digits) || VIETNAM_LANDLINE_RE.test(digits)) return digits;
  return '';
}

export function extractPhones(text: string): string[] {
  const seen = new Set<string>();
  const phones: string[] = [];
  for (const match of text.matchAll(PHONE_RE)) {
    const phone = normalizePhone(match[0]);
    if (phone && !seen.has(phone)) {
      seen.add(phone);
      phones.push(phone);
    }
  }
  return phones;
}

export function phonesForComment(row: { phone?: string; phones?: string[] }): string[] {
  if (row.phones?.length) return row.phones;
  if (row.phone) return [row.phone];
  return [];
}
